templates/NDArray.kt

```diff
@@ -90,6 +90,53 @@ interface NDArray<T> {
      */
     fun toMutableList() = MutableList(size) { getGeneric(it) }
 
+    /**
+     * Determine whether every element of one NDArray is sufficiently close to the corresponding element
+     * of another array.  Two elements are considered close if
+     *
+     * abs(ele1 - ele2) < (atol + rtol * abs(ele1))
+     *
+     * @param other   the array to compare this one to
+     * @param rtol    the maximum relative (i.e. fractional) difference to allow between elements
+     * @param atol    the maximum absolute difference to allow between elements
+     */
+    fun <R> allClose(other: NDArray<R>, rtol: Double=1e-05, atol: Double=1e-08): Boolean {
+        if (!(shape().toIntArray() contentEquals other.shape().toIntArray()))
+            return false
+        for (i in 0 until size) {
+            val a = getDouble(i)
```

Comment:
getDouble is only guaranteed to be implemented by `NDArray<Double>` backends, so this function is only guaranteed to work with `Double` currently. The current `Default` implementations do cast between primitives, but I'm not sure if that will be the case for all backends (future NDArray implementations have been requested, e.g. #70). If one ends up needing to call e.g. `getDouble` in context where the type is only known to be generic `NDArray<T>`, a try/catch is probably necessary to make sure the user doesnt get a undiagnosable error (this currently gives them the error `Exception in thread "main" java.lang.IllegalStateException: Double methods not implemented for generic NDArray`, which is difficult to decipher for a user)

Note that the goal of the `get$PRIMITIVE` methods is to have a non-boxing accessor that should only be called by code that knows the primitive type is correct. In particular, they were designed to be called by the extension functions with a known primitive type, such as `fun NDArray<Double>.foo(...) = this.getDouble()`, thus retaining type safety for the user. Ideally these `get$PRIMITIVE` methods would be marked `internal` so users wouldn't be able to see them at all, but there's some issues with module access that haven't been resolved yet (it looks like they are resolved in the new mpp though, so perhaps #77 will allow us to).